### PR TITLE
Add support for Nostr Zaps

### PR DESCRIPTION
**Pull request by mutatrum**

### CODE REVIEW

**Comment by mutatrum** [comment]
Pending https://github.com/blc-org/una/pull/37.

**Comment by 0xsommerfeld** [comment]
Does this work if I run the branch or do I need to wait for the issues you mentioned to get fixed as well?

**Comment by mutatrum** [comment]
I ran into a problem because of `gc-canceled-invoices-on-the-fly=true`. If you don't have that set in your LND config, I think it defaults to false and it should be ok to run then, but I have not tested this.

**Comment by ok300** [comment]
Each new zap request will create a zap request

https://github.com/Dolu89/ligess/blob/216d0e2f14fc65a1d34141a41e66e20c01bfc515/nostr.js#L76-L78

which the server will remember (i.e. store in RAM) until it either expires or is paid.

What is preventing someone from DDoS-ing via an endless stream of zap requests?

**Comment by mutatrum** [comment]
Good point, although this is already a DDoS vector for any lightning address handler. I don't see a rate limit option, as one can generate a new pubkey for every request and make valid zap requests.

A simple way of at least preventing OOM would be to define a max number of entries and remove them FIFO.

**Comment by seaKyou1** [comment]
Hey, thanks for your work, this is truly amazing! Got it the zaps to work.

To the other rookies out there, these are some steps I had to do additionally to the README.md instructions (other than checking out the specific pull request (ID=9) from github):

1. before running "yarn install" I actually had to install yarn first:
  sudo echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list
  sudo apt-get update
  sudo apt-get install yarn

2. I also had to install the nodemon module for npm:
  sudo npm install -g nodemon

3. I personally then also had to change the port of this app. If you have issues you can change it in this file: router.js

Remember to use HEX (!) values for the macaroons and nostr private key.


Again, thank you so much, this is fun!

**Comment by ok300** [comment]
@mutatrum yes, configuring a "max number of pending zaps" limit can defend against OOM.

But ironically that enables another kind of DoS attack, whereby the attacker floods with enough requests to reach that limit. At that point, nobody else can zap to this receiver.

Is there a way to defend against that?

There's a similar thread in the BTCPay repo: https://github.com/btcpayserver/btcpayserver/discussions/4642#discussioncomment-5085307

**Comment by mutatrum** [comment]
> 

I used npm for the install, and that was a straightforward `npm install`. Can't remember doing anything else to be honest. I haven't used the yarn installation method.

**Comment by mutatrum** [comment]
> Is there a way to defend against that?

Not that I can think of, other than rate-limiting the HTTP requests, the rest is trivial to generate. Either way at some point the service will be unavailable during the attack, but that's the nature of open internet I guess.

Thanks for the thread link, added a comment there.

**Comment by 0xsommerfeld** [comment]
>     3. I personally then also had to change the port of this app. If you have issues you can change it in this file: router.js

Looking at the relevant code:

https://github.com/Dolu89/ligess/blob/d8c71626a403a5d40f6f88b134506d7776a34d3a/router.js#L75

You should be able to change the binding port by passing the environment variable `PORT` in the `.env` file
E.g.:

```bash
PORT=<some other port other than the default(3000)>
```

**Comment by Sebastix** [comment]
Can confirm the zaps are working here, it's running now for two days. Some people had some errors but I think it's due their used clients. Still debugging with them but nothing has come out relevant for this integration of zaps.

**Comment by mutatrum** [comment]
Some clients are buggy. I just had a zap request without any relays, so it didn't send the zap note.

**Comment by 0xsommerfeld** [comment]
It's working for me: note17pfxx953dym0lzlf5flxu6ztdnjgdfqsuk83ghu7f9fmxv0vkmvq0dglaf
Thanks @mutatrum !

File: package.json [hunk 1]
```diff
@@ -13,7 +13,8 @@
     "bech32": "^2.0.0",
     "dotenv": "^10.0.0",
     "fastify": "^3.20.2",
-    "una-wrapper": "^0.0.7-beta"
+    "una-wrapper": "^0.0.10-beta",
```
Comment (Dolu89) [inline package.json:16]: I deployed a new version of Una with your changes, can you change it to `^0.0.11-beta`?

**Review by Dolu89 — CHANGES_REQUESTED** [review]

**Comment by mutatrum** [comment]
Should it return an error if the zap request fails any of the verifications? Currently it logs the issue, but still creates the invoice. Problem with this is that it leads to silent failures of zaps not working. Upside is that even with buggy implementations, payment will work, just not the zap.

**Comment by mutatrum** [comment]
All dependencies have been bumped. I also changed the error handling to deny any invalid zap requests. Running this version now, to see if anything weird pops up.

**Review by Dolu89 — APPROVED** [review]

**Pull request merged**